### DISCUSSION TITLE
Block bootstrap confidence intervals

### DISCUSSION THREAD

**Discussion by RogerHar** [dtcenter/METplus#1659] [Statistical Computation] [answered]
The last paragraph of [Appendix D Confidence Intervals](https://met.readthedocs.io/en/latest/Users_Guide/appendixD.html) of the MET User Guide mentions concerns accounting for dependency and says that "One method that is particularly appropriate for serially dependent data is the circular block resampling procedure for step 2". Am I correct to think this is implemented in METviewer only?

I have found the flag `<circular_block_bootstrap>` that turns on/off the circular block bootstrap in the [documentation for the METViewer Batch Plotting Module](https://metviewer.readthedocs.io/en/latest/Users_Guide/batch.html). (As it's **true** by default I assume it's always used when _not_ using batch plotting.) But how is the block size chosen? Is it decided automatically, and if so, by what method?

Thanks,
Roger

**Comment by TatianaBurek** [comment, marked as answer]
Roger,
Circular block resampling procedure is implemented in METcalcpy and is used by METviewer.
METviewer turns on Circular block bootstrap by default for all its modes - web and batch - unless it is specified in XML as `<circular_block_bootstrap>false</circular_block_bootstrap>`
You can find the implementation of CBB [here](https://github.com/dtcenter/METcalcpy/blob/78c6c9045fd95fdc55a1cc3421284e589a9be80e/metcalcpy/bootstrap.py#L469)
The block length for CBB should be longer than the length of dependence in the data, but much shorter than the size of the data. Generally, the square root of the sample size is a good choice. And that is how METplotpy calculates it : [example](https://github.com/dtcenter/METcalcpy/blob/78c6c9045fd95fdc55a1cc3421284e589a9be80e/metcalcpy/agg_stat.py#L709)

**Reply by RogerHar** [reply]
Thanks for explaining that Tatiana. Personally I'd favour including an option to allow the user to set a block length and over-ride the default in agg_stat.

I've seen the suggestion to use the square root of the sample size in a couple of Eric Gilleland's papers but I'm not clear where it originally came from and I've not seen it elsewhere - other software I've come across (the [b.star function in the R package 'np'](https://rdrr.io/cran/np/man/b.star.html)` and the [bootstrap.optimal_block_length function of the python package 'arch'](https://arch.readthedocs.io/en/latest/bootstrap/generated/arch.bootstrap.optimal_block_length.html)) uses the method of [Politis and White (2004)](https://doi.org/10.1081/ETC-120028836) which gives a block size proportional to the _cube_ root of the sample size, and also depends on the autocorrelation function.

**Comment by TatianaBurek** [comment]
I created a new issue to include a new parameter to METviewer config files.
https://github.com/dtcenter/METviewer/issues/399

**Reply by RogerHar** [reply]
Sorry but I can't see how that issue about Revision series relates to bootstrap block size??

**Reply by TatianaBurek** [reply]
I accidentally linked  incorrect issue to my comment. Should be fixed now.
